src/azure-cli-core/azure/cli/core/command_recommender.py

```diff
@@ -0,0 +1,191 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import difflib
+
+import azure.cli.core.telemetry as telemetry
+from knack.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class CommandRecommender():
+    """Recommend a command for user when user's command fails.
+    It combines alladin recommendations and examples in help files."""
+
+    def __init__(self, command, parameters, extension, cli_ctx):
+        """
+        :param command: The command name in user's input, which should be corrected if misspelled.
+        :type command: str
+        :param parameters: The parameter arguments in users input.
+        :type parameters: list
+        :param extension: The extension name in user's input if the command comes from an extension.
+        :type extension: str
+        :param cli_ctx: CLI context when parser fails.
+        :type cli_ctx: knack.cli.CLI
+        """
+        self.command = command.strip()
+        self.extension = extension
+        self.cli_ctx = cli_ctx
+
+        self.parameters = self._normalize_parameters(parameters)
+        self.help_examples = []
+        self.aladdin_recommendations = []
+
+    def set_help_examples(self, examples):
+        """Set recommendations from help files"""
+
+        self.help_examples.extend(examples)
+
+    def _set_aladdin_recommendations(self):
+        """Set recommendations from aladdin service.
+        Call the aladdin service API, parse the response and set the recommendations.
+        """
+
+        import json
+        import requests
+        from requests import RequestException
+        from http import HTTPStatus
+        from azure.cli.core import __version__ as version
+
+        api_url = 'https://app.aladdin.microsoft.com/api/v1.0/suggestions'
+        headers = {'Content-Type': 'application/json'}
```

Comment:
I posted this in another comment, but we should have a way to disable Aladdin calls for a period of time if it encounters repeated errors. The scenarios I'm thinking about are:

1. A user in an air-gapped cloud will never be able to hit the service (we are working on a way to download the model and query it locally, but that won't involve the service itself), so they should not waste cycles trying to hit an endpoint that does not exist for them.
2. A user experiencing a temporary internet outage will want this to work once they reconnect, so any disabling should have a time limit to prevent the feature from being disabled forever.

Reply:
Yes, air gapped cloud is a good case here, we need consider to remove their requests for the api.